Add WiFi GATT service

ramonhpr opened the pull request:
This patch adds a GATT service to set up the WiFi which the gateway will connect with.

ramonhpr committed:
ble: gatt: Add wifi service

This patch adds a GATT service to enable clients to set the SSID
to the desired wifi network to connect with.

Signed-off-by: Ramon Ribeiro <rhpr@cesar.org.br>

ramonhpr committed:
ble: gatt: Add password characteristic

This patch adds a password characteristic to wifi GATT service so
it is possible to authenticate with the network.

Signed-off-by: Ramon Ribeiro <rhpr@cesar.org.br>

vitbaq commented on netsetup/srv_gatt_wifi.py:
```diff
@@ -0,0 +1,62 @@
+
+    def __init__(self, bus, index, service):
+        Characteristic.__init__(self, bus, index, self.PSWD_UUID,
+                                ["encrypt-write"], service)
```
Use "write" only.
Currently, the knot app setup is not doing any security process.

vitbaq reviewed (changes requested):
Remove encrypt on pswd characteristic. In this first moment, KNoT setup mobile App is not doing any security process.
Also, it's suggested adding a characteristic that passes a wifi status (e.g. if the gw is connected or not). What do you think?

ramonhpr commented:
@vitbaq I agree with the characteristic status. But I will add it when I make the integration with connman, because it already has a dbus property like that.

ramonhpr committed:
ble: gatt: Allow value update watch

This patch change the default method WriteValue in class Characteristic
allowing the signal propertiesChanged to be called. So it is possible
to monitor the value changed by the GATT service.

Signed-off-by: Ramon Ribeiro <rhpr@cesar.org.br>

ramonhpr committed:
connman: Add initial support

This patch adds initial support to connect to connman manager
interface.

ConnMan project provides a daemon for managing internet connections
within embedded devices running the Linux operating system.

Signed-off-by: Ramon Ribeiro <rhpr@cesar.org.br>

ramonhpr committed:
connman: Register wifi agent to allow connection

This patch registers a connman agent to send the password to a wifi
network when it is requested to the user.

For more information about connman agent API see the following link:
https://git.kernel.org/pub/scm/network/connman/connman.git/tree/doc/agent-api.txt

Signed-off-by: Ramon Ribeiro <rhpr@cesar.org.br>

ramonhpr committed:
connman: Add scan_wifi method

This patch adds the Scan method to connman client. As described by connman
technology API, the Scan is called by an timeout to avoid the blocking
the mainloop. And a signal ServicesChanged is emitted when the scan is
finished.

See:
https://git.kernel.org/pub/scm/network/connman/connman.git/tree/doc/technology-api.txt

Signed-off-by: Ramon Ribeiro <rhpr@cesar.org.br>

ramonhpr committed:
connman: Add connect to Wi-Fi support

This patch adds a method to connect to a simple network with SSID and
password. Once the Connect method is called by the manager interface,
the method RequestInput from WifiAgent is called and the password is
set to the following service.

Signed-off-by: Ramon Ribeiro <rhpr@cesar.org.br>

ramonhpr committed:
ble: gatt: wifi: Connect when characteristics set

This patch makes a connection with the wifi network set by the
characteristics SSID and Password in the GATT service be set.

Signed-off-by: Ramon Ribeiro <rhpr@cesar.org.br>

vitbaq reviewed (changes requested):
Plz, add the GATT wifi service on the BLE advertisement (line 40, ble.py). If possible, don't use the UUID value directly, use from ser_gatt_wifi.py.

ramonhpr committed:
ble: Remove hard coded service UUID

This patch passes KNoT GATT application to KNoT advertisement to get
its services and UUIDs.

Signed-off-by: Ramon Ribeiro <rhpr@cesar.org.br>

ramonhpr committed:
ble: advertisement: Toggle service uuid

This patch toggles the service UUIDs in advertisement packet

Signed-off-by: Ramon Ribeiro <rhpr@cesar.org.br>

jcorde commented on netsetup/ble_util/characteristic.py:
```diff
@@ -8,6 +8,7 @@
 import logging
```
About the commit message for 71ed67a87764754c7015d95d4f4031908aa49c7d

jcorde replied:
Please, update the title for something more descriptive. Something like `ble: gatt: Allow value update watch` would be more fit.

jcorde replied:
Typo: 
> the signal propertiesChanged **to** be called

jcorde replied:
Please, revert:
>  changed by the **service GATT**

jcorde commented on netsetup/ble_util/characteristic.py:
```diff
@@ -67,8 +68,13 @@ def ReadValue(self, options):
+        if 'write' not in self.flags:
+            raise NotSupportedException()
+        if not value:
+            logging.error('Invalid value length')
```
Maybe something like `No value received` would be more appropriated

jcorde commented on netsetup/ble.py:
```diff
@@ -34,9 +34,11 @@ def __init__(self, bus):
 
```
About commit message for b207a3820d740bf30ef12fa93b5abf9ac2963b46

jcorde replied:
Please, use something like:
> Add `gatt_knot` field to store application being advertised.

jcorde commented on netsetup/ble.py:
```diff
@@ -34,9 +34,11 @@ def __init__(self, bus):
-        self.add_service_uuid("a8a9e49c-aa9a-d441-9bec-817bb4900e30")
+        self.gatt_knot = gatt_knot
+        self.add_service_uuid(gatt_knot.services[0].UUID)
+        self.current_srv_gatt = 0
```
What is this for?

jcorde commented on netsetup/ble.py:
```diff
@@ -12,6 +12,7 @@
 from dbus.service import method as dbus_method
```
About message for commit 0da7f37960b281803871831b11ba9612f5642bd4

jcorde replied:
> patch **adds** a GATT

jcorde commented on netsetup/srv_gatt_wifi.py:
```diff
@@ -0,0 +1,53 @@
+
+    def __on_ssid_changed(self, interface, changed, invalidated):
+        self.ssid = changed['Value']
+        logging.info('SSID changed %s' % convert_array_to_str(self.ssid))
```
Please, use `SSID changed *to* %s`

jcorde commented on netsetup/ble.py:
```diff
@@ -44,6 +47,13 @@ def __init__(self, bus, index, gatt_knot, ad_name):
         self.include_tx_power = True
 
+    def toogle_uuid(self):
+        qtd_srv = len(self.gatt_knot.services)
```
**qtd** sorry yo no hablo portugues

jcorde commented on netsetup/ble.py:
```diff
@@ -44,6 +47,13 @@ def __init__(self, bus, index, gatt_knot, ad_name):
+        qtd_srv = len(self.gatt_knot.services)
+        self.current_srv_gatt = (self.current_srv_gatt + 1) % qtd_srv
+        uuid = self.gatt_knot.services[self.current_srv_gatt].UUID
+        self.service_uuids = []
```
This should be called right after
```
self.ad_manager.UnregisterAdvertisement(
            self.ad_knot,
            reply_handler=lambda: None,
            error_handler=on_error)
```

jcorde commented on netsetup/ble.py:
```diff
@@ -44,6 +47,13 @@ def __init__(self, bus, index, gatt_knot, ad_name):
         self.add_local_name(ad_name)
         self.include_tx_power = True
 
+    def toogle_uuid(self):
```
as this does not remove the old UUID, I see this as a `set_next_uuid()`  or something similar

jcorde commented on netsetup/__main__.py:
```diff
@@ -87,20 +87,16 @@ def main():
 
```
About message for commit c2fb91536142c734a511282c434c4ef14feeaa2c

jcorde replied:
This patch toggles the service UUIDs in advertisement packet

jcorde commented on netsetup/connman.py:
```diff
@@ -0,0 +1,28 @@
+#!/usr/bin/env python
```
About message for commit cecea97f1d5f4d6e1962fa640337e0c22f656d63

Please, explain what connman does.

Also, use a more descriptive title. Something like: Add initial connection manager support.

jcorde commented on netsetup/connman.py:
```diff
@@ -0,0 +1,28 @@
+            logging.error('%s: Connman is probably not running.' % err)
+            return
+
+    def __init__(self):
```
Please, place `__init__` as the first function of the class

jcorde commented on netsetup/connman.py:
```diff
@@ -0,0 +1,28 @@
+            return dbus.Interface(
+                self.bus.get_object(CONNMAN_SERVICE_NAME, '/'),
+                CONNMAN_MANAGER_INTERFACE)
+        except DBusException as err:  # ignore if connmand is not running
```
This exception is not only for connman errors. Please, update the comment

jcorde commented on netsetup/connman.py:
```diff
@@ -10,6 +10,32 @@
 
```
About message for commit f77f2be6be97cfaff3c09e8e5a0d0f2f583139da

jcorde replied:
This patch **registers** a connman agent to send the password to a wifi
network when **it** is requested to **the** user.

For more **information** about connman agent API see the following link:
https://git.kernel.org/pub/scm/network/connman/connman.git/tree/doc/agent-api.txt

jcorde commented on netsetup/connman.py:
```diff
@@ -0,0 +1,124 @@
+    def RequestInput(self, path, fields):
+        response = {}
+        if 'Passphrase' in fields:
+            response.update({'Passphrase': self.passphrase})
```
Where did you define `self.passphrase`?

jcorde replied:
Did you have to?

jcorde commented on netsetup/connman.py:
```diff
@@ -54,3 +55,44 @@ def __init__(self):
+        properties = iface.GetProperties()
+        if not properties.get('Powered'):
+            iface.SetProperty('Powered', dbus.Boolean(True))
+            logging.info('Wifi enabled')
```
You're sending the command to enable the Wi-Fi, but no verification is done. Please, update the logging to something like `'Enabling wifi'` or add a confirmation step before it.

jcorde commented on netsetup/connman.py:
```diff
@@ -54,3 +55,44 @@ def __init__(self):
+
+    def __get_wifi_technology_path(self):
+        if not self.manager:
+            logging.error('Unable to get wifi technology. Connman not running')
```
This check should be done at the constructor.

jcorde commented on netsetup/connman.py:
```diff
@@ -54,3 +55,44 @@ def __init__(self):
+        services = []
+
+        if not path:
+            logging.error('Not found technology wifi')
```
Shouldn't this check be done inside `__get_wifi_technology_path()`?

jcorde commented on netsetup/connman.py:
```diff
@@ -54,3 +55,44 @@ def __init__(self):
+
+        wifi_tech = dbus.Interface(
+            self.bus.get_object(CONNMAN_SERVICE_NAME, path),
+            CONNMAN_TECHNOLOGY_INTERFACE)
```
Why not including 
```
dbus.Interface(
            self.bus.get_object(CONNMAN_SERVICE_NAME, path),
            CONNMAN_TECHNOLOGY_INTERFACE)
```
inside `__get_wifi_technology_path()` and changing its name to `__get_wifi_technology()`?

jcorde commented on netsetup/connman.py:
```diff
@@ -54,3 +55,44 @@ def __init__(self):
         self.agent = WifiAgent(self.bus, WIFI_AGENT_PATH)
         self.manager.RegisterAgent(WIFI_AGENT_PATH)
+
+    def __enable_wifi(self, iface):
```
Why calling it `iface` here, but receiving a technology?

jcorde commented on netsetup/connman.py:
```diff
@@ -54,3 +55,44 @@ def __init__(self):
+
+        for obj_path, properties in self.manager.GetTechnologies():
+                if properties.get('Type') == 'wifi':
+                        return obj_path
```
Shouldn't a warning be printed in case of not finding wifi technology?

jcorde commented on netsetup/connman.py:
```diff
@@ -54,3 +55,44 @@ def __init__(self):
+
+        def on_services_changed(currentServices, rmServices):
+            self.signal_match.remove()
+            on_scan_completed(currentServices)
```
Please, add a blank line after the function definition

jcorde commented on netsetup/connman.py:
```diff
@@ -11,6 +11,7 @@
 CONNMAN_SERVICE_NAME = 'net.connman'
```
About message for commit 7b6690bccb858bba2858e1f80e8fbb3767671afa:

jcorde replied:
Maybe the title could be more descriptive as `connman: Add connect to Wi-Fi support`

jcorde replied:
>This patch adds a method to connect **to** a simple network with SSID and
password.

jcorde commented on netsetup/connman.py:
```diff
@@ -96,3 +97,28 @@ def on_services_changed(currentServices, rmServices):
         wifi_tech.Scan(reply_handler=lambda: logging.info('Scan complete'),
                        error_handler=lambda err: logging.error(err))
+
+    def __get_wifi_service_path(self, name):
```
Why not including `__get_wifi_service_path(self, name)` as a inner function of `__do_connect(self, path)`, and change it to `__do_connect(self, name)`?

jcorde commented on netsetup/connman.py:
```diff
@@ -96,3 +97,28 @@ def on_services_changed(currentServices, rmServices):
+            logging.error('Unable to get wifi technology. Connman not running')
+            return
+
+        for obj_path, properties in self.manager.GetServices():
```
Please, log in case of not finding the target obj_path

jcorde commented on netsetup/connman.py:
```diff
@@ -96,3 +97,28 @@ def on_services_changed(currentServices, rmServices):
+        service.Connect(reply_handler=lambda: logging.info('Connected'),
+                        error_handler=lambda err: logging.error(err))
+
+    def connect_wifi(self, name, password):
```
This is redundant with `__do_connect()`
Shouldn't this be something like `scan_and_connect_wifi()`?
